src/ruby/pb/test/xds_client.rb

```diff
@@ -109,29 +164,62 @@ def get_client_stats(req, _call)
       num_failures: watcher['no_remote_peer'] + watcher['rpcs_needed']
     );
   end
+
+  def get_client_accumulated_stats(req, _call)
+    LoadBalancerAccumulatedStatsResponse.new(
+      num_rpcs_started_by_method: $num_rpcs_started_by_method,
+      num_rpcs_succeeded_by_method: $num_rpcs_succeeded_by_method,
+      num_rpcs_failed_by_method: $num_rpcs_failed_by_method
+    )
+  end
 end
 
 # execute 1 RPC and return remote hostname
-def execute_rpc(op, fail_on_failed_rpcs)
+def execute_rpc(op, fail_on_failed_rpcs, rpc_stats_key)
   remote_peer = ""
   begin
     op.execute
     if op.metadata.key?('hostname')
       remote_peer = op.metadata['hostname']
     end
   rescue GRPC::BadStatus => e
-    GRPC.logger.info("ruby xds: rpc failed:|#{e.message}|, " \
-                     "this may or may not be expected")
     if fail_on_failed_rpcs
       raise e
     end
   end
+  if remote_peer.empty?
+    $num_rpcs_failed_by_method[rpc_stats_key] += 1
+  else
+    $num_rpcs_succeeded_by_method[rpc_stats_key] += 1
+  end
   remote_peer
 end
 
+def execute_rpc_in_thread(op, rpc_stats_key)
+  num_open_threads = $keep_open_threads.size
+  if num_open_threads % 50 == 0
+    GRPC.logger.info("number of keep_open_threads = #{num_open_threads}")
+  end
+  $keep_open_threads << Thread.new {
```

Comment:
does `$keep_open_threads` need to be global?

can we change this method to return the new thread, and have `keep_open_threads` be a local of `run_test_loop`?

That way, it's easier to reason about how this is operation thread safe

Reply:
Comment addressed. Now `keep_open_threads` is local to `run_test_loop`.